src/bluesearch/database/topic_rule.py

```diff
@@ -0,0 +1,125 @@
+# Blue Brain Search is a text mining toolbox focused on scientific use cases.
+#
+# Copyright (C) 2020  Blue Brain Project, EPFL.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+"""Implementation of the TopicRule data structure."""
+from __future__ import annotations
+
+import re
+from typing import Iterable, Optional
+
+import pydantic
+
+from bluesearch.database.article import ArticleSource
+from bluesearch.database.topic_info import TopicInfo
+
+
+class TopicRule(pydantic.BaseModel, arbitrary_types_allowed=True):
+    """The topic rule."""
+
+    # None always represent wildcards
+    level: Optional[str] = None  # "article" or "journal"
+    source: Optional[ArticleSource] = None  # "arxiv", ... , "pubmed"
+    pattern: Optional[re.Pattern] = None  # regex pattern to match
+
+    @pydantic.validator("level")
+    def check_level_value(cls, value):
+        """Check the level parameter value."""
+        if value is not None and value not in {"article", "journal"}:
+            raise ValueError(f"Unsupported level {value}")
+        return value
+
+    @pydantic.validator("source", pre=True)
+    def convert_source_to_article_source(cls, value) -> ArticleSource | None:
+        """Check the source parameter value."""
+        if value is not None:
+            try:
+                value = ArticleSource(value)
+            except ValueError:
+                raise ValueError(f"Unsupported source {value}") from None
+
+        return value
+
+    @pydantic.validator("pattern", pre=True)
+    def convert_pattern_to_re(cls, value) -> re.Pattern | None:
+        """Check the pattern parameter value."""
+        if value is not None:
+            try:
+                value = re.compile(value)
+            except re.error:
+                raise ValueError(f"Unsupported pattern {value}") from None
+
+        return value
+
+    def match(self, topic_info: TopicInfo) -> bool:
+        """Determine whether a topic_info matches the rule."""
+        # Source
+        if self.source is not None and self.source is not topic_info.source:
+            return False
+
+        if self.pattern is None:
+            return True
+
+        if self.level is None or self.level == "article":
+            for topic_list in topic_info.article_topics.values():
```

Comment:
Regarding 1), I do see the potential utility, however, as of right now each source has only one "topic kind".

IMO there multiple solutions

a) No action (just "concatenated" all topics from different "topic kind") and do the matching
b) Use `list[str]` instead of `dict[str, list[str]]` and make an assumption that each source has a single "topic kind"
c) Keep multiple "topic kind" logic (`dict[str, list[str]]`), however, add an extra attribute to `TopicRule` (e.g. `topic_kind`) so that one can actually use it in filtering

Reply:
I guess I am fine with `a)`, i.e. the solution you have implemented, since for the moment you are right that there is a 1-to-1 relationship between `topic kind` per `source`.

But then I think we need to clearly document the fact that `topic kind` info is completely ignored.

Reply:
Added some extra documentation here: 5df7005